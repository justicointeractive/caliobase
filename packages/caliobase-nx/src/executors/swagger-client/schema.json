{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "title": "SwaggerClient executor",
  "description": "",
  "type": "object",
  "properties": {
    "input": {
      "type": "string",
      "description": "path to an openapi/swagger spec file"
    },
    "output": {
      "type": "string",
      "description": "path to the directory where the generated client will be saved"
    },
    "generateSpecTarget": {
      "type": "string",
      "description": "a target to be executed that should generate a swagger file"
    },
    "dotenv": {
      "type": "string",
      "description": "path to a dotenv file to load vars from"
    }
  },
  "required": []
}
