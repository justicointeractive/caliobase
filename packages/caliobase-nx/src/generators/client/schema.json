{
  "$schema": "http://json-schema.org/schema",
  "$id": "Client",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "apiProjectName": {
      "type": "string",
      "description": "",
      "x-prompt": "Which project provides the backend api?",
      "x-dropdown": "projects"
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the project (used for linting)",
      "alias": "t"
    },
    "directory": {
      "type": "string",
      "description": "A directory where the project is placed",
      "alias": "d"
    }
  },
  "required": [
    "name",
    "apiProjectName"
  ]
}
