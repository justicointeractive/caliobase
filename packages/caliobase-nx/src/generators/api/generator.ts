import {
  formatFiles,
  generateFiles,
  getWorkspaceLayout,
  names,
  Tree,
} from '@nrwl/devkit';
import { applicationGenerator } from '@nrwl/nest/src/generators/application/application';
import * as path from 'path';
import { ApiGeneratorSchema } from './schema';

interface NormalizedSchema extends ApiGeneratorSchema {
  projectName: string;
  projectRoot: string;
  projectDirectory: string;
  parsedTags: string[];
}

function normalizeOptions(
  tree: Tree,
  options: ApiGeneratorSchema
): NormalizedSchema {
  const name = names(options.name).fileName;
  const projectDirectory = options.directory
    ? `${names(options.directory).fileName}/${name}`
    : name;
  const projectName = projectDirectory.replace(new RegExp('/', 'g'), '-');
  const projectRoot = `${getWorkspaceLayout(tree).appsDir}/${projectDirectory}`;
  const parsedTags = options.tags
    ? options.tags.split(',').map((s) => s.trim())
    : [];

  return {
    ...options,
    projectName,
    projectRoot,
    projectDirectory,
    parsedTags,
  };
}

export default async function (tree: Tree, options: ApiGeneratorSchema) {
  const normalizedOptions = normalizeOptions(tree, options);
  await applicationGenerator(tree, { name: options.name });
  generateFiles(
    tree,
    path.join(__dirname, 'files'),
    normalizedOptions.projectRoot,
    {
      cmsProjectName: null,
      ...options,
      ...normalizedOptions,
      tmpl: '',
    }
  );
  await formatFiles(tree);
}
