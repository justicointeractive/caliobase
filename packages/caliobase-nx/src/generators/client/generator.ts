import {
  formatFiles,
  generateFiles,
  getWorkspaceLayout,
  names,
  offsetFromRoot,
  Tree,
} from '@nx/devkit';
import { libraryGenerator } from '@nx/js/src/generators/library/library';
import * as path from 'path';
import { modifyProjectConfiguration } from '../../lib/modifyProjectConfiguration';
import { ClientGeneratorSchema } from './schema';

interface NormalizedSchema extends ClientGeneratorSchema {
  projectName: string;
  projectRoot: string;
  projectDirectory: string;
  parsedTags: string[];
}

function normalizeOptions(
  tree: Tree,
  options: ClientGeneratorSchema
): NormalizedSchema {
  const name = names(options.name).fileName;
  const projectDirectory = options.directory
    ? `${names(options.directory).fileName}/${name}`
    : name;
  const projectName = projectDirectory.replace(new RegExp('/', 'g'), '-');
  const projectRoot = `${getWorkspaceLayout(tree).libsDir}/${projectDirectory}`;
  const parsedTags = options.tags
    ? options.tags.split(',').map((s) => s.trim())
    : [];

  return {
    ...options,
    projectName,
    projectRoot,
    projectDirectory,
    parsedTags,
  };
}

function addFiles(tree: Tree, options: NormalizedSchema) {
  const templateOptions = {
    ...options,
    ...names(options.name),
    offsetFromRoot: offsetFromRoot(options.projectRoot),
    tmpl: '',
  };
  generateFiles(
    tree,
    path.join(__dirname, 'files'),
    options.projectRoot,
    templateOptions
  );
}

export default async function (tree: Tree, options: ClientGeneratorSchema) {
  const normalizedOptions = normalizeOptions(tree, options);
  await libraryGenerator(tree, {
    name: options.name,
    directory: options.directory,
    skipFormat: true,
  });
  addFiles(tree, normalizedOptions);
  modifyProjectConfiguration(tree, options.name, (config) => {
    config.targets = {
      ...config.targets,
      'generate-sources': {
        executor: '@caliobase/caliobase-nx:swagger-client',
        outputs: [`${normalizedOptions.projectRoot}/src`],
        options: {
          input: `./${normalizedOptions.projectRoot}/assets/openapi.json`,
          output: `./${normalizedOptions.projectRoot}/src/lib`,
          name: `${normalizedOptions.projectName}.ts`,
        },
        dependsOn: [{ projects: 'dependencies', target: 'swagger' }],
      },
    };
    config.implicitDependencies = [
      ...(config.implicitDependencies ?? []),
      options.apiProjectName,
    ];

    return config;
  });
  modifyProjectConfiguration(
    tree,
    options.apiProjectName,
    (apiProjectConfig) => {
      apiProjectConfig.targets = {
        ...apiProjectConfig.targets,
        swagger: {
          executor: 'nx:run-commands',
          outputs: [`${normalizedOptions.projectRoot}/assets`],
          dependsOn: [
            {
              target: 'build',
              projects: 'self',
            },
          ],
          options: {
            command: `node ./dist/${apiProjectConfig.root}/main.js --write-swagger-and-exit=./${normalizedOptions.projectRoot}/assets/openapi.json`,
          },
        },
      };

      return apiProjectConfig;
    }
  );

  // TODO: update nx.json's targetDefaults build dependsOn "^generate-sources", "generate-sources"

  await formatFiles(tree);
}
