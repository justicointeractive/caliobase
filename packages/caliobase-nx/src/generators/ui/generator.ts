import {
  formatFiles,
  generateFiles,
  GeneratorCallback,
  getWorkspaceLayout,
  names,
  offsetFromRoot,
  Tree,
} from '@nrwl/devkit';
import { Linter } from '@nrwl/linter';
import { applicationGenerator } from '@nrwl/react';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';
import * as path from 'path';
import { modifyProjectConfiguration } from '../../lib/modifyProjectConfiguration';
import { addDependencyVersionsToPackageJson } from '../../lib/versions';
import { UiGeneratorSchema } from './schema';
import assert = require('assert');

interface NormalizedSchema extends UiGeneratorSchema {
  projectName: string;
  projectRoot: string;
  projectDirectory: string;
  parsedTags: string[];
}

function normalizeOptions(
  tree: Tree,
  options: UiGeneratorSchema
): NormalizedSchema {
  const name = names(options.name).fileName;
  const projectDirectory = options.directory
    ? `${names(options.directory).fileName}/${name}`
    : name;
  const projectName = projectDirectory.replace(new RegExp('/', 'g'), '-');
  const projectRoot = `${getWorkspaceLayout(tree).appsDir}/${projectDirectory}`;
  const parsedTags = options.tags
    ? options.tags.split(',').map((s) => s.trim())
    : [];

  return {
    ...options,
    projectName,
    projectRoot,
    projectDirectory,
    parsedTags,
  };
}

function addFiles(tree: Tree, options: NormalizedSchema) {
  const templateOptions = {
    ...options,
    ...names(options.name),
    offsetFromRoot: offsetFromRoot(options.projectRoot),
    template: '',
  };
  generateFiles(
    tree,
    path.join(__dirname, 'files'),
    options.projectRoot,
    templateOptions
  );
}

export default async function (tree: Tree, options: UiGeneratorSchema) {
  const tasks: GeneratorCallback[] = [];
  const normalizedOptions = normalizeOptions(tree, options);
  await applicationGenerator(tree, {
    ...options,
    e2eTestRunner: 'cypress',
    linter: Linter.EsLint,
    unitTestRunner: 'jest',
    skipFormat: true,
    style: 'css',
  });
  addFiles(tree, normalizedOptions);
  modifyProjectConfiguration(tree, options.name, (config) => {
    assert(config.targets);
    config.targets['serve'].options[
      'proxyConfig'
    ] = `${normalizedOptions.projectRoot}/proxy.conf.json`;
    return config;
  });
  await formatFiles(tree);

  tasks.push(
    addDependencyVersionsToPackageJson(tree, ['@caliobase/caliobase-ui'])
  );

  return runTasksInSerial(...tasks);
}
