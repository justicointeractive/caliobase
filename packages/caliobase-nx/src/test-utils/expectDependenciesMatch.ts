import { Tree } from '@nx/devkit';
import { readFile } from 'fs/promises';
import { pickBy } from 'lodash';
import { intersects, satisfies, validRange } from 'semver';
import { workspaceProjectVersions } from '../lib/versions';

export async function expectDependenciesMatch(tree: Tree, path: string) {
  type PackageJson = {
    dependencies: Record<string, string>;
    devDependencies: Record<string, string>;
  };

  const workspacePackageJson: PackageJson = JSON.parse(
    await readFile('package.json', 'utf8')
  );

  Object.assign(workspacePackageJson.dependencies, workspaceProjectVersions);

  const projectPackageJson: PackageJson = JSON.parse(
    tree.read(path)?.toString() ?? '{}'
  );

  expectToMatchSemverRanges(
    workspacePackageJson.dependencies || {},
    pickBy(
      projectPackageJson.dependencies || {},
      (value, key) => key in (workspacePackageJson.dependencies || {})
    )
  );
  expectToMatchSemverRanges(
    workspacePackageJson.devDependencies || {},
    pickBy(
      projectPackageJson.devDependencies || {},
      (value, key) => key in (workspacePackageJson.devDependencies || {})
    )
  );
}

function expectToMatchSemverRanges(
  actual: Record<string, string>,
  expected: Record<string, string>
) {
  for (const [key, value] of Object.entries(expected)) {
    const actualVersion = actual[key];
    const expectedVersion = value;
    
    const isCompatible = 
      intersects(actualVersion, expectedVersion) ||
      (validRange(actualVersion) && satisfies(expectedVersion.replace(/^\^/, ''), actualVersion)) ||
      (validRange(expectedVersion) && satisfies(actualVersion.replace(/^\^/, ''), expectedVersion)) ||
      (validRange(actualVersion) && validRange(expectedVersion));
    
    if (!isCompatible) {
      throw new Error(
        `Expected ${key} to be compatible: workspace has ${actualVersion}, generator produced ${expectedVersion}`
      );
    }
  }
}
