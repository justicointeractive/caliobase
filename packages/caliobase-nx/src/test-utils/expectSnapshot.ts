import { Tree } from '@nrwl/devkit';

export function expectSnapshot(tree: Tree, path: string) {
  expect(tree.read(path)?.toString()).toMatchSnapshot(path);
}
