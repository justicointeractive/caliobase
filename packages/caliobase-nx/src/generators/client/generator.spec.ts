import { readProjectConfiguration, Tree } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { expectSnapshot } from '../../test-utils/expectSnapshot';
import apiGenerator from '../api/generator';
import generator from './generator';
import { ClientGeneratorSchema } from './schema';

describe('client generator', () => {
  let appTree: Tree;
  const options: ClientGeneratorSchema = {
    name: 'client',
    apiProjectName: 'api',
  };

  beforeEach(async () => {
    appTree = createTreeWithEmptyWorkspace();
    await apiGenerator(appTree, { name: 'api' });
  });

  it('should run successfully', async () => {
    await generator(appTree, options);
    const config = readProjectConfiguration(appTree, 'client');
    expect(config).toBeDefined();
    expectSnapshot(appTree, 'package.json', [
      [/"@caliobase\/caliobase": "[0-9.]+"/g, '"@caliobase/caliobase": "*"'],
    ]);
    expectSnapshot(appTree, 'apps/api/src/main.ts');
    expectSnapshot(appTree, 'libs/client/project.json');
    expectSnapshot(appTree, 'libs/client/src/index.ts');
    expectSnapshot(appTree, 'libs/client/src/lib/client.ts');
  });
});
