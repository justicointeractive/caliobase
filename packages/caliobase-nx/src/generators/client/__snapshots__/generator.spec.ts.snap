// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client generator should run successfully: apps/api/src/main.ts 1`] = `
"import { CaliobaseModule } from '@caliobase/caliobase';
import { Logger } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { AppModule } from './app/app.module';

async function bootstrap() {
  const app = await NestFactory.create(AppModule.forRootAsync(), {
    cors: true,
  });

  const globalPrefix = 'api';
  app.setGlobalPrefix(globalPrefix);

  await CaliobaseModule.bootstrap(app, {
    migration: {
      migrationsDir: './apps/apps-api/migrations',
    },
  });

  const port = process.env.PORT || 3324;
  await app.listen(port);
  Logger.log(
    \`ðŸš€ Application is running on: http://localhost:\${port}/\${globalPrefix}\`
  );
}

bootstrap().catch((err) => {
  console.error(err);
  process.exit(1);
});
"
`;

exports[`client generator should run successfully: libs/client/project.json 1`] = `
"{
  "name": "libs-client",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/client/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "build": {
      "executor": "@nx/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/libs/client",
        "main": "libs/client/src/index.ts",
        "tsConfig": "libs/client/tsconfig.lib.json",
        "assets": ["libs/client/*.md"]
      }
    },
    "lint": {
      "executor": "@nx/eslint:lint"
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/client/jest.config.ts"
      }
    },
    "generate-sources": {
      "executor": "@caliobase/caliobase-nx:swagger-client",
      "outputs": ["./libs/client/src"],
      "options": {
        "input": "././libs/client/assets/openapi.json",
        "output": "././libs/client/src/lib",
        "name": "libs-client.ts"
      },
      "dependsOn": [
        {
          "projects": "dependencies",
          "target": "swagger"
        }
      ]
    }
  },
  "implicitDependencies": ["apps-api"]
}
"
`;

exports[`client generator should run successfully: libs/client/src/index.ts 1`] = `
"export * from './lib/libs-client';
"
`;

exports[`client generator should run successfully: libs/client/src/lib/libs-client.ts 1`] = `
"// this will be replaced with a generated file when the generate-sources target is executed
"
`;

exports[`client generator should run successfully: package.json 1`] = `
"{
  "name": "@proj/source",
  "dependencies": {
    "@aws-sdk/client-ses": "^3.171.0",
    "@caliobase/caliobase": "*",
    "@nestjs/common": "^11.0.0",
    "@nestjs/core": "^11.0.0",
    "@nestjs/platform-express": "^11.0.0",
    "@nestjs/swagger": "7.3.0",
    "axios": "^1.6.0",
    "nodemailer": "^6.7.5",
    "reflect-metadata": "^0.1.13",
    "rxjs": "^7.8.0",
    "tslib": "^2.3.0"
  },
  "devDependencies": {
    "@nestjs/schematics": "11.0.5",
    "@nestjs/testing": "^11.0.0",
    "@nx/eslint": "21.2.3",
    "@nx/eslint-plugin": "21.2.3",
    "@nx/jest": "21.2.3",
    "@nx/js": "21.2.3",
    "@nx/nest": "21.2.3",
    "@nx/node": "21.2.3",
    "@nx/web": "21.2.3",
    "@nx/webpack": "21.2.3",
    "@swc-node/register": "~1.9.1",
    "@swc/core": "~1.5.7",
    "@swc/helpers": "~0.5.11",
    "@types/jest": "^29.5.12",
    "@types/node": "~18.16.9",
    "@types/nodemailer": "^6.4.4",
    "@typescript-eslint/eslint-plugin": "^7.16.0",
    "@typescript-eslint/parser": "^7.16.0",
    "eslint": "~8.57.0",
    "eslint-config-prettier": "^10.0.0",
    "jest": "^29.7.0",
    "jest-environment-node": "^29.7.0",
    "jsonc-eslint-parser": "^2.1.0",
    "prettier": "^2.6.2",
    "ts-jest": "^29.1.0",
    "ts-node": "10.9.1",
    "typescript": "~5.8.2"
  }
}
"
`;
